staging/src/k8s.io/endpointslice/topologyheuristics/manager.go

```diff
@@ -0,0 +1,344 @@
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.activeHeuristicsByService[serviceKey]; !ok {
+		m.activeHeuristicsByService[serviceKey] = make(map[discoveryv1.AddressType]Heuristic)
+	}
+
+	m.activeHeuristicsByService[serviceKey][addrType] = heuristic
+}
+
+// desiredHeuristicForService returns the Heuristic corresponding to the
+// topology annotation in the service.
+//   - enabled return value signifies that the topology annotation determines an
+//     enabled value.
+//   - exists return value signifies existence of the heuristic in the heuristic
+//     manager.
+//
+// The Heuristic returned represents the "desired state", which could be
+// different from the "current state". Use activeHeuristicForService() to get
+// the "current state".
+func (m *Manager) desiredHeuristicForService(service *corev1.Service) (topology Heuristic, enabled, exists bool) {
+	desiredHeuristicName, desiredHeuristicEnabled := HeuristicNameFromAnnotations(service.GetAnnotations())
+	if !desiredHeuristicEnabled {
+		return nil, false, false
+	}
+
+	// Special case for backward compatibility: If the topology annotation value
+	// is set to "Auto", we need to return defaultHeuristic.
+	if desiredHeuristicName == "Auto" || desiredHeuristicName == "auto" {
+		heuristic, ok := m.heuristicsByName[m.defaultHeuristicName]
+		return heuristic, true, ok
+	}
+
+	for heuristicName, heuristic := range m.heuristicsByName {
+		if desiredHeuristicName == heuristicName {
+			return heuristic, true, true
+		}
+	}
+
+	return nil, true, false
+}
+
+// RecognizableHeuristic returns true if the given heuristicName is recognized
+// by the Manager. A topology is recognized if it is:
+//   - Supported by the Manager.
+//   - (or)
+//   - Present in nonEPSHintHeuristicNames (i.e. not implemented through
+//     EndpointSlice hints)
+func (m *Manager) RecognizableHeuristic(heuristicName string) bool {
+	var recognizableHeuristicNames []string
+	recognizableHeuristicNames = append(recognizableHeuristicNames, "Auto", "auto")
+	recognizableHeuristicNames = append(recognizableHeuristicNames, m.nonEPSHintHeuristicNames...)
+	for name := range m.heuristicsByName {
+		recognizableHeuristicNames = append(recognizableHeuristicNames, name)
+	}
+	return slices.Contains(recognizableHeuristicNames, heuristicName)
+}
+
+// HeuristicNameFromAnnotations function returns the value of the topology
+// annotation, if one is present. It also identifies whether the annotation
+// value indicates that the topology aware routing is enabled or disabled.
+//
+// The function uses two annotation keys to determine the topology annotation
+// value:
+//   - v1.AnnotationTopologyMode
+//   - v1.DeprecatedAnnotationTopologyAwareHints.
+//
+// If both annotation keys are set, v1.DeprecatedAnnotationTopologyAwareHints
+// has precedence.
+func HeuristicNameFromAnnotations(annotations map[string]string) (heuristicName string, enabled bool) {
```

Comment:
I don't understand what is the purpose of the bool, `HeuristicNameFromAnnotations` should return just the heuristicName string and the caller implement the handling of the names and its special cases